fix: Set the gid of mounted /data with fsGroup

kimwnasptd opened the pull request:
Closes https://github.com/canonical/charmed-kubeflow-uats/issues/163

## Changes
1. Uses `fsGroup` in the PodSpec
2. Updates integration tests to call `kubectl` directly
## Review Notes
~~The `fsGroup` is also respected in MicroK8s.~~ This was **NOT true**! PTAL at https://github.com/canonical/minio-operator/pull/207#issuecomment-2717275949

For the test I tried to make it future proof by capturing the group of `/data` and ensuring the running process belongs to that group. To avoid hardcoding specific group numbers.

[This article](https://discourse.charmhub.io/t/updated-podspec-yaml-new-features/2124) was super helpful in helping understand how to set `fsGroup`.

DnPlas reviewed (commented):
Thanks @kimwnasptd, the changes make sense but after testing, they give me a slightly different result than what you posted [here](https://github.com/canonical/charmed-kubeflow-uats/issues/163#issuecomment-2713870186).

For starters, I get this message every time I log into the container:

```
kubectl exec -ti -nminio-163 minio-0 -- /bin/bash                                                            
Defaulted container "minio" out of: minio, juju-pod-init (init)                                                                        
groups: cannot find name for group ID 2000    
```

Also, the ownership of `/data/*` is entirely for `_daemon_`, rather than `_daemon_` and the `2000` group:

```
_daemon_@minio-0:/$ ls -la /data/                                                                                                      
total 12                                                                                                                               
drwxrwxrwx 3 root     root     4096 Mar 12 05:27 .                                                                                     
drwxr-xr-x 1 root     root     4096 Mar 12 05:27 ..
drwxr-xr-x 6 _daemon_ _daemon_ 4096 Mar 12 05:27 .minio.sys
```

And `root` is the owner of `/data`:

```
_daemon_@minio-0:/$ ls -lh /                                                                                                           
total 68K                                                                                                                              
lrwxrwxrwx   1 root root    7 Jan 26 02:05 bin -> usr/bin                                                                              
drwxr-xr-x   2 root root 4.0K Apr 18  2022 boot                                                                                        
drwxrwxrwx   3 root root 4.0K Mar 12 05:27 data   
```
I used `juju deploy minio --channel latest/edge/pr-207 --revision 429 --trust` for deploying the charm. Can you just confirm that we are actually getting the expected output?

kimwnasptd commented:
> Thanks @kimwnasptd, the changes make sense but after testing, they give me a slightly different result than what you posted https://github.com/canonical/charmed-kubeflow-uats/issues/163#issuecomment-2713870186.

Good catch! I'm assuming you tested with MicroK8s, and I also reproduced with MicroK8s the logs you showed.

After also [chatting](https://chat.canonical.com/canonical/pl/xompmhbg4tff9yo8peww7ds6nc) with @ktsakalozos it looks like MicroK8s:
1. Will not handle `fsGroup`, for hostpath PVCs at least
2. Will always change the permissions of PVC mounts, of hostpath-storage, to `drwxrwxrwx`. Thus the container processes will always have permissions there (and they don't do `chown` at all)

BUT, this means the tests passing was a bug. Specifically, because the [test was doing](https://github.com/canonical/minio-operator/pull/207/files#diff-08291d7f4ad7b2caecacb6ae26af540ebbaede9fe42d5e8e1b39d2bdefe4c2d4R259-R275):
```python
    kubectl_cmd = (
        "kubectl",
        "exec",
        f"--namespace={ops_test.model_name}",
        "-ti",
        f"{APP_NAME}-0",
        "-c",
        APP_NAME,
        "--",
        "id",
        "-G",
    )

    ret_code, groups, _ = await ops_test.run(*kubectl_cmd)
    assert ret_code == 0
  

    # groups will be ["584792 2000"]
    assert fs_group in groups. # <-- here's the bug
```

So the test would check if string `0` (group of `/data`) is in the string `584792 2000`. Which is true, because `0` is inside `2000`... But if we'd change the check to `assert fs_group in groups.split(" ")` then the test would fail.
```
AssertionError: assert '0\r\n' in ['584792', '2000\r\n']
 +  where ['584792', '2000\r\n'] = <built-in method split of str object at 0x7c88375fc1f0>(' ')
 +    where <built-in method split of str object at 0x7c88375fc1f0> = '584792 2000\r\n'.split
```

The problem here is the fact that microK8s is not respecting `fsGroup` (my PR message was wrong in this case, edited), thus we can't have a test that tests this behavor. 

My proposal then is to fully remove the test in this case, since we rely on MicroK8s for testing. WDYT @DnPlas?

kimwnasptd commented:
> Also, when deploying the minio ckf-1.9/stable, we actually get some logs from the process:

This looks to be an issue when we integrated the rock. By refreshing the charm to use the old upstream image, I see the logs that you posted
```bash
juju refresh minio \
    --resource oci-image=minio/minio:RELEASE.2021-09-03T03-56-13Z \
    --path ./minio_ubuntu-20.04-amd64.charm
```

I also see some other differences in the OCI images when trying to run them locally. Created https://github.com/canonical/minio-rock/issues/7 to track the logs and the execution of the image.

kimwnasptd commented:
Note that the logs I posted in https://github.com/canonical/charmed-kubeflow-uats/issues/163#issuecomment-2713870186 are for EKS, in which the `fsGroup` behavior is as expected and the issue with permissions is resolved.

DnPlas commented:
Unfortunately, this fix does not guarantee that after upgrading the charm from `ckf-1.9` to `latest/edge` (eventually `ckf-1.10/stable`) the issue is fixed. I have left a comment (https://github.com/canonical/charmed-kubeflow-uats/issues/163#issuecomment-2719621472) explaining how I found out this and a potential workaround.

kimwnasptd commented:
Closing the PR for now, and once we catch a breath from the release we can re-evaluate our strategy.
